### PR TITLE
Start on colwise blog post

### PR DESCRIPTION
**Pull request by hadley**

### CODE REVIEW

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
+```
+
+This post is the latest in a series of post leading up the the dplyr 1.0.0 release. So far, the series has covered:
+
+* [Major lifecycle changes](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-is-coming-soon/).
+* [New `summarise()` features](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-summarise/).
+* [`select()`, `rename()`, `relocate()`](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-select-rename-relocate/).
+
+Today, I wanted to talk a little bit about the new `across()` function that makes it easy to perform the same operation to multiple columns.
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:25]:
```suggestion
Today, I wanted to talk a little bit about the new `across()` function that makes it easy to perform the same operation on multiple columns.
```

Comment (batpigandme) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:25]: I think [perform takes on](https://inspirassion.com/en/prep/perform) as its preposition here, though you could switch "perform" to _do_ and then keep "to" instead of _on_.

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+If you're interested in living life on the edge (or trying out anything you see in this blog post), you can install the development version of dplyr with:
+
+```{r, eval = FALSE}
+devtools::install_github("tidyverse/dplyr")
+```
+
+Note that the development version won't become 1.0.0 until it's released, but it has all the same features.
+
+```{r setup}
+library(vctrs)
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:38]: Do you want to say anything re. why vctrs is being attached as well?

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+It's often useful to perform the same operation on multiple columns, but copying and pasting is both tedious and error prone:
+
+```{r, eval = FALSE}
+df %>% 
+  group_by(g1, g2) %>% 
+  summarise(a = mean(a), b = mean(b), c = mean(c), d = mean(c))
+```
+
+You can now rewrite such code to use `across()`, which lets you apply a transformation to multiple variables selected with the same syntax a [`select()` and `rename()`]( [tidy selection](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-select-rename-relocate/#select-and-renaming)):
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:52]:
```suggestion
You can now rewrite such code using `across()`, which lets you apply a transformation to multiple variables selected with the same syntax as [`select()` and `rename()`]( [tidy selection](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-select-rename-relocate/#select-and-renaming)):
```

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  group_by(g1, g2) %>% 
+  summarise(across(a:d, mean))
+
+# or 
+df %>% 
+  group_by(g1, g2) %>% 
+  summarise(across(is.numeric, mean))
+```
+
+You might be familiar with `summarise_if()` and `summarise_at()` that we previously recommended for this sort of operation. Later in the blog post we'll come back to why we now prefer `across()`. But for now, let's dive into the basics of `across()`.
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:65]:
```suggestion
You might be familiar with `summarise_if()` and `summarise_at()` which we previously recommended for this sort of operation. Later in the blog post we'll come back to why we now prefer `across()`. But for now, let's dive into the basics of `across()`.
```

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+## Basic usage
+
+`across()` has two primary arguments:
+
+* The first argument, `.cols`, selects the columns you want to operate on.
+  It uses the tidy select syntax so you can pick columns by position, name,
+  function of name, type, or any combination.
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:73]:
```suggestion
  function of name, type, or any combination thereof using Boolean operators.
```

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+* The second argument, `.fns`, is a function or list of functions to apply to
+  each column. You can use also purrr style formulas like `~ .x / 2`. 
+
+Here are a couple of examples of `across()` in conjunction with its favourite verb, `summarise()`:
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:78]: This is a little cutesy…

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+* You can use it with [multiple summary functions](https://dplyr.tidyverse.org/dev/articles/colwise.html#multiple-functions).
+
+* You can use it with [any dplyr verb](https://dplyr.tidyverse.org/dev/articles/colwise.html#other-verbs).
+
+* If needed, you can access the name of the column currently being processed
+  with [`cur_column()`](https://dplyr.tidyverse.org/dev/articles/colwise.html#current-column).
+
+## Why `across()`? 
+
+If you've tackled this problem with an older version of dplyr, you might've used one of the functions had an `_if`, `_at`, or `_all` suffix. These functions solved a pressing need and are used by many people, but are now superseded. This means that they'll stay around, but will only recieve critical bug fixes.
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:107]:
```suggestion
If you've tackled this problem with an older version of dplyr, you might've used one of the functions with an `_if`, `_at`, or `_all` suffix. These functions solved a pressing need and are used by many people, but are now superseded. This means that they'll stay around, but will only receive critical bug fixes. 
```

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  with [`cur_column()`](https://dplyr.tidyverse.org/dev/articles/colwise.html#current-column).
+
+## Why `across()`? 
+
+If you've tackled this problem with an older version of dplyr, you might've used one of the functions had an `_if`, `_at`, or `_all` suffix. These functions solved a pressing need and are used by many people, but are now superseded. This means that they'll stay around, but will only recieve critical bug fixes. 
+
+Why did we decide to move away from these functions in favour of `across()`?
+
+1.  `across()` makes it possible to compute useful summaries that were 
+    previously impossible. For example, it's now trivial to summarise
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:112]:
```suggestion
    previously impossible. For example, it's now easy to summarise
```

Comment (batpigandme) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:112]: Easy isn't great, but I loathe trivial.

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+4.  `across()` doesn't need `vars()`. The `_at()` functions are the only place 
+    in dplyr where you have to use `vars()`, which makes them unusual, 
+    and hence harder to learn and remember.
+
+Why did it take it long to discover `across()`? Surprisingly, the key idea that makes `across()` works came out of our low-level work on the [vctrs](http://vctrs.r-lib.org/) package, where we learnt that you can have a column of a data frame that is itself a data frame. It's a bummer that we had a few false starts before we discovered `across()`, but even with hindsight, I don't see how we could've skipped the intermediate steps.
+
+## How do you convert existing code?
+
+Fortunately, if you have code that uses the `_if`, `_at`, or `_all()` functions, it's generally straightforward to update:
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:143]:
```suggestion
If you have code that uses the `_if`, `_at`, or `_all()` functions, it's generally straightforward to update:
```

Comment (batpigandme) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:143]: To me, "Fortunately" makes it sound more like updating might be strictly necessary (which it isn't, given what we mean by superseded). I think it's good to be extra careful here, since there's still a tenuous understanding of the term out there.

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+df %>% mutate_at(vars(c(x, starts_with("y"))), mean, na.rm = TRUE)
+# ->
+df %>% mutate(across(c(x, starts_with("y")), mean, na.rm = TRUE))
+
+df %>% mutate_all(mean, na.rm = TRUE)
+# ->
+df %>% mutate(across(everything(), mean, na.rm = TRUE))
+```
+
+If you've used multiple `_if`/`_at`/`_all` functions in a row, you should also consider if it's now possible to collapse into a single call, using the new features of `across()`.
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:171]:
```suggestion
If you've used multiple `_if`/`_at`/`_all` functions in a row, you should also consider if it's now possible to collapse them into a single call, using the new features of `across()`.
```

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 9 above)
Comment (batpigandme) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:143]: Oh, I see you have extra reassurance later, but I still think we can axe fortunately.

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+Again, you don't need to worry about these functions going away in the short-term, but it's good practice to update your code to use our latest recommendations.
```
Comment (batpigandme) [inline content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:173]: > but it's good practice to update your code to use our latest recommendations.

Is there another way of saying this that alludes to the merits of `across()` rather than our recommendations?

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 2 above)
Comment (hadley) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:38]: No, because that's a mistake 😬

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 6 above)
Comment (hadley) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:78]: "in conjunction with the verb with which it pairs most naturally" is a bit wordy. Any  ideas?

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 9 above)
Comment (hadley) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:143]: Yeah, I'm struggled with this — I want to imply that it's easy, but you don't have to do it.  Maybe: If you want to update your existing code ...

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 11 above)
Comment (hadley) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:173]: "use the latest features"?

That reminds me that I need to mention that `across()` is slower currently but we have a planning for improving performance in 1.1.0

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 6 above)
Comment (hadley) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:78]: Oh I can just make it "Here are a couple of examples of `across()` used with `summarise()`: "

Comment (batpigandme) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:78]: Yes!

File: content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown (hunk 11 above)
Comment (batpigandme) [reply content/blog/2020-04-dplyr-1-0-0-colwise/index.Rmarkdown:173]: I like it.
